### ISSUE TITLE
Enhancement: [naming-convention] Check non-native types

### ISSUE THREAD

**Issue by Betree** [typescript-eslint/typescript-eslint#6854]
### Before You File a Proposal Please Confirm You Have Done The Following...

- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22enhancement%3A+plugin+rule+option%22) and found none that match my proposal.
- [X] I have searched the [current rule list](https://typescript-eslint.io/rules/#supported-rules) and found no rules that match my proposal.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### My proposal is suitable for this project

- [X] I believe my proposal would be useful to the broader TypeScript community (meaning it is not a niche proposal).

### Link to the rule's documentation

https://typescript-eslint.io/rules/naming-convention/

### Description

Using the `naming-convention` rule, we would like to enforce rules on custom types. Our practical use case is to prefix all `GraphQLEnumType` and `GraphQLObjectType` instances with "GraphQL".

The rule could look like that:

```json
"@typescript-eslint/naming-convention": [
  "error",
  {
    "selector": "variable",
    "types": ["GraphQLEnumType", "GraphQLObjectType"],
    "prefix": ["GraphQL"]
  }
]
```

### Fail

```typescript
const Expense = new GraphQLObjectType({ ... })
```

### Pass

```typescript
const GraphQLExpense = new GraphQLObjectType({ ... })
```

**Comment by JoshuaKGoldberg** [comment]
Hmm, this is interesting. I've definitely seen a lot of versions of this preference over the years: variables assigned to particular types being named in particular ways.

In theory we could bring in the format discussed in #6017 & implemented in #4436 to the naming-convention rule. We could do something like allowing `{ "selector": "type", "type": { "from": ... }, "format": ... }`.

However, the `naming-convention` rule is already pretty complex (>700 lines!) and hard to deal with. Each time we add to it, it gets harder to maintain. Given that this is the only instance of this request I could find in the first 3 pages of feature requests to the rule, let's mark it as `evaluating community engagement`. If lots of folks ask for this, it'll be worth the effort.

**Comment by bradzacher** [comment]
This issue has been open for 3 months and has not received any community engagement.
As such - we're taking this as signal that this is not important to the broader community, and so we are rejecting this proposal. Thanks for filing!
